### PR TITLE
Implement primitives necessary to do EK-based credential activation

### PR DESCRIPTION
**Pull request by twitchy-jsonp**
This depends on #56, so ignore all the `AuthCommand` stuff.

This PR implements:

 * `PolicySecret` - which is necessary given the default EK template mandates policy auth.
 * `ActivateCredentialUsingAuth` - which allows the caller to specify authorization parameters (in my case, so I can use session auth to authorize `HandleEndorsement`).

### CODE REVIEW

File: tpm2/tpm2.go [hunk 1]
```diff
@@ -621,6 +619,32 @@ func PolicyPassword(rw io.ReadWriter, handle tpmutil.Handle) error {
...
+	return concat(handles, auth, params)
+}
+
+// PolicySecret sets a secret authorization requirement on the provided entity.
+func PolicySecret(rw io.ReadWriter, entityHandle tpmutil.Handle, entityPassword string, policyHandle tpmutil.Handle, policyNonce, cpHash, policyRef []byte) error {
+	cmd, err := encodePolicySecret(entityHandle, entityPassword, policyHandle, policyNonce, cpHash, policyRef)
+	if err != nil {
+		return err
+	}
+	_, err = runCommand(rw, TagSessions, cmdPolicySecret, tpmutil.RawBytes(cmd))
```
Comment (awly) [inline tpm2/tpm2.go]: Parse and return the ticket from response

File: tpm2/tpm2.go [hunk 2]
```diff
@@ -808,7 +832,20 @@ func decodeActivateCredential(in []byte) ([]byte, error) {
...
 func ActivateCredential(rw io.ReadWriter, activeHandle, keyHandle tpmutil.Handle, activePassword, protectorPassword string, credBlob, secret []byte) ([]byte, error) {
-	cmd, err := encodeActivateCredential(activeHandle, keyHandle, activePassword, protectorPassword, credBlob, secret)
+	return ActivateCredentialUsingAuth(rw, []AuthCommand{
+		{Session: HandlePasswordSession, Attributes: AttrContinueSession, Auth: []byte(activePassword)},
+		{Session: HandlePasswordSession, Attributes: AttrContinueSession, Auth: []byte(protectorPassword)},
+	}, activeHandle, keyHandle, credBlob, secret)
+}
+
+// ActivateCredentialUsingAuth associates an object with a credential, using the
+// given set of authorizations. Returns decrypted certificate information.
```
Comment (awly) [inline tpm2/tpm2.go]: clarify that len(auth) must be exactly 2

File: tpm2/tpm2.go [hunk 3]
```diff
@@ -621,6 +619,32 @@ func PolicyPassword(rw io.ReadWriter, handle tpmutil.Handle) error {
...
+	}
+	params, err := tpmutil.Pack(policyNonce, cpHash, policyRef, int32(0))
+	if err != nil {
+		return nil, err
+	}
+	return concat(handles, auth, params)
+}
+
+// PolicySecret sets a secret authorization requirement on the provided entity.
+func PolicySecret(rw io.ReadWriter, entityHandle tpmutil.Handle, entityPassword string, policyHandle tpmutil.Handle, policyNonce, cpHash, policyRef []byte) error {
```
Comment (awly) [inline tpm2/tpm2.go]: Take `AuthCommand` instead of a string for `entityPassword`

File: tpm2/tpm2.go [hunk 4]
```diff
@@ -808,7 +832,20 @@ func decodeActivateCredential(in []byte) ([]byte, error) {
...
-	cmd, err := encodeActivateCredential(activeHandle, keyHandle, activePassword, protectorPassword, credBlob, secret)
+	return ActivateCredentialUsingAuth(rw, []AuthCommand{
+		{Session: HandlePasswordSession, Attributes: AttrContinueSession, Auth: []byte(activePassword)},
+		{Session: HandlePasswordSession, Attributes: AttrContinueSession, Auth: []byte(protectorPassword)},
+	}, activeHandle, keyHandle, credBlob, secret)
+}
+
+// ActivateCredentialUsingAuth associates an object with a credential, using the
+// given set of authorizations. Returns decrypted certificate information.
+func ActivateCredentialUsingAuth(rw io.ReadWriter, auth []AuthCommand, activeHandle, keyHandle tpmutil.Handle, credBlob, secret []byte) ([]byte, error) {
```
Comment (awly) [inline tpm2/tpm2.go]: Add a test for this func, including a case where auth validation fails

**Comment by twitchy-jsonp** [comment]
sshfs is broken so I'll have to dance between computers (via commits) for testing. I'll let you know when everything is implemented/working on my side (dont merge as commits come in).

**Comment by twitchy-jsonp** [comment]
PTAL! :unicorn:

File: tpm2/tpm2_test.go [hunk 5]
```diff
@@ -203,16 +203,38 @@ func TestCombinedEndorsementTest(t *testing.T) {
...
 	}
+
+	recoveredCredential2, err := ActivateCredentialUsingAuth(rw, []AuthCommand{
+		{Session: HandlePasswordSession, Attributes: AttrContinueSession, Auth: []byte(defaultPassword)},
+		{Session: HandlePasswordSession, Attributes: AttrContinueSession, Auth: []byte(emptyPassword)},
+	}, keyHandle, parentHandle, credBlob, encryptedSecret0)
+	if err != nil {
+		t.Fatalf("ActivateCredentialWithAuth failed: %v", err)
+	}
+	if bytes.Compare(credential, recoveredCredential2) != 0 {
```
Comment (awly) [inline tpm2/tpm2_test.go]: `bytes.Equal`

File: tpm2/tpm2_test.go [hunk 6]
```diff
@@ -203,16 +203,38 @@ func TestCombinedEndorsementTest(t *testing.T) {
...
+	}
+
+	_, err = ActivateCredentialUsingAuth(rw, []AuthCommand{
+		{Session: HandlePasswordSession, Attributes: AttrContinueSession, Auth: []byte("wrong password")},
+		{Session: HandlePasswordSession, Attributes: AttrContinueSession, Auth: []byte(emptyPassword)},
+	}, keyHandle, parentHandle, credBlob, encryptedSecret0)
+	if err == nil {
+		t.Fatal("ActivateCredentialUsingAuth: error == nil, expected response status 0x98e (authorization failure)")
+	}
+	if err.Error() != "response status 0x98e" {
```
Comment (awly) [inline tpm2/tpm2_test.go]: This is fragile. Use `strings.Contains`

File: tpm2/tpm2.go [hunk 7]
```diff
@@ -635,6 +635,44 @@ func PolicyPassword(rw io.ReadWriter, handle tpmutil.Handle) error {
...
+		return nil, err
+	}
+	params, err := tpmutil.Pack(policyNonce, cpHash, policyRef, expiry)
+	if err != nil {
+		return nil, err
+	}
+	return concat(handles, auth, params)
+}
+
+// PolicySecret sets a secret authorization requirement on the provided entity.
```
Comment (awly) [inline tpm2/tpm2.go]: Mention that `expiry` is in seconds

File: tpm2/tpm2_test.go [hunk 8]
```diff
@@ -203,16 +203,38 @@ func TestCombinedEndorsementTest(t *testing.T) {
...
+		t.Errorf("ActivateCredentialUsingAuth: error = %v, expected response status 0x98e (authorization failure)", err)
+	}
```
Comment (awly) [inline tpm2/tpm2_test.go]: Add a test case with one AuthCommand
```
        ActivateCredentialUsingAuth(rw, []AuthCommand{
		{Session: HandlePasswordSession, Attributes: AttrContinueSession, Auth: []byte(defaultPassword)},
	}, keyHandle, parentHandle, credBlob, encryptedSecret0)
```

File: tpm2/tpm2_test.go [hunk 9]
```diff
@@ -203,16 +203,38 @@ func TestCombinedEndorsementTest(t *testing.T) {
...
+
+	recoveredCredential2, err := ActivateCredentialUsingAuth(rw, []AuthCommand{
+		{Session: HandlePasswordSession, Attributes: AttrContinueSession, Auth: []byte(defaultPassword)},
+		{Session: HandlePasswordSession, Attributes: AttrContinueSession, Auth: []byte(emptyPassword)},
+	}, keyHandle, parentHandle, credBlob, encryptedSecret0)
+	if err != nil {
+		t.Fatalf("ActivateCredentialWithAuth failed: %v", err)
+	}
+	if bytes.Compare(credential, recoveredCredential2) != 0 {
+		t.Fatalf("Credential and recovered credential differ: got %v, want %v", recoveredCredential2, credential)
```
Comment (awly) [inline tpm2/tpm2_test.go]: Errorf

**Comment by twitchy-jsonp** [comment]
Doneskies. PTAL!

**Pull request merged**